feat: optional command timeout field in config file that defaults  to syncTimeout

marshallford opened the pull request:
Adds the ability to set an optional timeout duration for all commands that defaults to the sync timeout instead of `time.Minute`. Attempts to resolve: #3049. Keep in mind that in order to fix the issue described, the sync timeout **must be** greater than or equal to the sum of the generator command timeouts. I am open to ideas or to scraping this completely. Thanks Flux team!

Example:

```yaml
version: 1
patchUpdated:
  generators:
  - command: kustomize build .
    timeout: 3m # if field is not provided, sync timeout is used
  patchFile: flux-patch.yaml
```

Couple of other notes:

1. ~~I looked into parsing `timeout` from the config yaml as the type `time.Duration` instead of a string but the use of the `ghodss/yaml` package made that a bit confusing (for a new golang user anyway). It would have been nice to return an error in `ParseConfigFile` if an invalid duration was given instead of using the syncTimeout as a default without notice.~~

Solved by C&P this snippet: https://stackoverflow.com/a/48051946

2. I'm not super happy with adding the field `defaultTimeout` to the `configAware` struct. Open to suggestions on how to improve passing through the sync timeout.

TODO:

- [ ] Investigate better way of passing around/through sync timeout flag
- [ ] Update docs once this method of setting a command timeout per generator is approved

marshallford commented:
@stefanprodan could I get a review on this? Would you be open to a schema change like this?

marshallford commented:
@stefanprodan ping

marshallford commented:
ping @hiddeco

rdubya16 commented:
@stefanprodan Any movement on this? This is blocking us as well.

mike-stewart commented:
@squaremo We're running into this bug as well. We are running a build of this PR in one of our clusters now, and it seems to be resolving the issue. Any chance this PR could be merged?

squaremo commented on pkg/daemon/daemon.go:
```diff
@@ -12,10 +12,10 @@ import (
-	"github.com/fluxcd/flux/pkg/api/v11"
-	"github.com/fluxcd/flux/pkg/api/v6"
-	"github.com/fluxcd/flux/pkg/api/v9"
+	v10 "github.com/fluxcd/flux/pkg/api/v10"
```
Was this (and the ff) changed by an IDE? I don't think it's necessary, or relevant to this PR.

marshallford replied:
Removed. VS Code + the Go extension was making this change on save.

squaremo commented on pkg/manifests/configfile.go:
```diff
@@ -84,6 +88,35 @@ func mustCompileConfigSchema() *jsonschema.Schema {
 	return schema
 }
 
+type Duration struct {
```
I think `metav1.Duration` (k8s.io/apimachinery/pkg/apis/meta/v1) could be used here.

marshallford replied:
I chose to create my own type so that I could marshal and unmarshal the duration.

squaremo replied:
`metav1.Duration` has its own marshalling code, so it parses a value of e.g., `5m`. If it's a requirement to allow floats here, I guess you are stuck with implementing your own. Is that a hard requirement?

marshallford replied:
Oh ok, perfect. I'm fine with no floats.

squaremo commented on pkg/manifests/configfile.go:
```diff
@@ -110,39 +143,29 @@ type ConfigFile struct {
 // Generator is an individual command for generating manifests.
-type Generator struct {
-	Command string `json:"command,omitempty"`
+type Command struct {
```
The preceding comment needs updating.

marshallford replied:
fixed.

squaremo commented on pkg/manifests/configfile.go:
```diff
@@ -427,52 +454,56 @@ func (cf *ConfigFile) execGenerators(ctx context.Context, generators []Generator
 		stdOutAndErr := bytes.NewBuffer(nil)
-		err := cf.execCommand(ctx, env, stdOutAndErr, stdOutAndErr, c)
+		timeout := c.Timeout
+		if &c.Timeout == nil || timeout.Duration < time.Second {
```
`if &c.Timeout == nil` does not make sense as a comparison. It says "is the _address_ of a field in a value `nil`", which will never be true. If you want the timeout value to be optional, make the type of the field a pointer.

marshallford replied:
I think I fixed this?

squaremo reviewed (changes requested):
Thanks for putting time into this! I appreciate the need, and I'm happy with the idea. Passing the default timeout is a bit awkward, as you mention; I'm fine with it though, at least it's obvious what's happening.
There's some bits in the implementation that need another look, which I've flagged up.

squaremo commented:
I've rebased this, and added a commit for using `metav1.Duration`. Once it passes CI I'll squash those early few commits and merge.

marshallford commented:
> I've rebased this, and added a commit for using `metav1.Duration`. Once it passes CI I'll squash those early few commits and merge.

Awesome Thanks!

squaremo commented:
@marshallford Can you do me a quick favour? We now [require a DCO](https://github.com/fluxcd/flux/pull/3228/checks?check_run_id=1309845179) (as of today!). It's pretty easy to apply it -- instructions are in that link, and you will need to git reset your local branch head first, to pick up my rebase.

squaremo commented:
Brill, thank you @marshallford!
